### PR TITLE
NEST-499: Bug fixing for Azure blob storage requests

### PR DESCRIPTION
**Pull request by wAsnk**
[NEST-499](https://lombiq.atlassian.net/browse/NEST-499)

[NEST-499]: https://lombiq.atlassian.net/browse/NEST-499?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Lombiq.Hosting.Azure.ApplicationInsights/Services/AzureBlobTelemetryFilter.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+﻿using Lombiq.Hosting.Azure.ApplicationInsights.Extensions;
+using Microsoft.ApplicationInsights.Channel;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ApplicationInsights.Extensibility;
+using System;
+
+namespace Lombiq.Hosting.Azure.ApplicationInsights.Services;
+
+public class AzureBlobTelemetryFilter : ITelemetryProcessor
```
Comment (Piedone) [inline Lombiq.Hosting.Azure.ApplicationInsights/Services/AzureBlobTelemetryFilter.cs:12]: Comment why this is necessary.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs [hunk 2]
```diff
@@ -76,19 +76,41 @@ public class ApplicationInsightsOptions
     public bool EnableOfflineOperation { get; set; }
 
     /// <summary>
-    /// Gets or sets <see cref="IgnoreFailureRegex"/> by compiling the given string into a regular expression.
+    /// Gets or sets <see cref="RequestIgnoreFailureRegex"/> by compiling the given string into a regular expression.
+    /// This will be used for RequestTelemetries.
```
Comment (Piedone) [inline Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs]: Use `<see>` when referring to types, also in the 3 other cases below.

File: Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs [hunk 3]
```diff
@@ -76,19 +76,41 @@ public class ApplicationInsightsOptions
...
-        set => IgnoreFailureRegex = new Regex(value, RegexOptions.Compiled, TimeSpan.FromSeconds(1));
+        get => RequestIgnoreFailureRegex?.ToString();
+        set => RequestIgnoreFailureRegex = new Regex(value, RegexOptions.Compiled, TimeSpan.FromSeconds(1));
     }
 
     /// <summary>
     /// Gets or sets a regular expression that will be used to set telemetry to success if it matches
-    /// <see cref="DependencyTelemetry.Data"/> or <see cref="RequestTelemetry.Url"/>. This is useful if
-    /// you have a lot of 404s or other errors that you don't want to see as failures in Application Insights.
+    /// <see cref="DependencyTelemetry.Data"/> or <see cref="RequestTelemetry.Url"/>. This is useful if you have a lot
```
Comment (Piedone) [inline Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs]: This comment mistakenly references `DependencyTelemetry.Data`, and the one on `DependencyIgnoreFailureRegex` `RequestTelemetry.Url`.

File: Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs [hunk 4]
```diff
@@ -76,19 +76,41 @@ public class ApplicationInsightsOptions
...
+    ///
     /// </summary>
-    public Regex IgnoreFailureRegex { get; set; }
+    public Regex RequestIgnoreFailureRegex { get; set; }
+
+    /// <summary>
+    /// Gets or sets <see cref="DependencyIgnoreFailureRegex"/> by compiling the given string into a regular expression.
+    /// This will be used for DependencyTelemetries.
+    /// </summary>
+    /// <example>You should use a regex pattern like "(?:\\/media\\/)".</example>
```
Comment (Piedone) [inline Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs]: Why is the capture group necessary?

File: Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs [hunk 5]
```diff
@@ -76,19 +76,41 @@ public class ApplicationInsightsOptions
     public bool EnableOfflineOperation { get; set; }
 
     /// <summary>
-    /// Gets or sets <see cref="IgnoreFailureRegex"/> by compiling the given string into a regular expression.
+    /// Gets or sets <see cref="RequestIgnoreFailureRegex"/> by compiling the given string into a regular expression.
+    /// This will be used for RequestTelemetries.
     /// </summary>
     /// <example>You should use a regex pattern like "(?:\\/favicon.ico$)|(?:\\/media\\/)".</example>
```
Comment (Piedone) [inline Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs]: Media is not necessary anymore, no?

File: Lombiq.Hosting.Azure.ApplicationInsights/Extensions/TelemetryExtensions.cs [hunk 6]
```diff
@@ -25,19 +25,22 @@ public static void SetAsIgnoredFailure(this OperationTelemetry operationTelemetr
...
     /// </summary>
     public static bool ShouldSetAsIgnoredFailure(this RequestTelemetry requestTelemetry, IServiceProvider serviceProvider) =>
-        ShouldSetAsIgnoredFailure(requestTelemetry.ResponseCode, requestTelemetry.Url.ToString(), serviceProvider);
+        IsResult400(requestTelemetry.ResponseCode) &&
+        serviceProvider.GetRequiredService<IOptions<ApplicationInsightsOptions>>().Value
+            .RequestIgnoreFailureRegex
+            ?.IsMatch(requestTelemetry.Url.ToString()) == true;
 
-    private static bool ShouldSetAsIgnoredFailure(string code, string data, IServiceProvider serviceProvider) =>
+    private static bool IsResult400(string code) =>
```
Comment (Piedone) [inline Lombiq.Hosting.Azure.ApplicationInsights/Extensions/TelemetryExtensions.cs]:
```suggestion
    private static bool IsResult4xx(string code) =>
```

**Review by Piedone — CHANGES_REQUESTED** [review]

File: Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs (hunk 4 above)
Comment (wAsnk) [reply Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs]: Because you asked for it :D

Comment (wAsnk) [reply Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs]: To be fair it is not necessary, the developer can decide if it is needed.

File: Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs (hunk 5 above)
Comment (wAsnk) [reply Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs]: It is just an example, if someone wants do to that why not?

Comment (wAsnk) [reply Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs]: Whatever, changed it.

File: Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs (hunk 4 above)
Comment (Piedone) [reply Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs]: But I didn't :). The group was necessary when using more patterns with `|`, but it's not necessary for matching a single one. I asked for the group to be a non-capturing one for performance then, but if groups are not needed then of course having a non-capturing one doesn't make sense either.

File: Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs (hunk 5 above)
Comment (Piedone) [reply Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs]: The examples should naturally be sensible. If you would not ignore media here (but only for dependencies) then we shouldn't have it here either.

File: Lombiq.Hosting.Azure.ApplicationInsights/Extensions/TelemetryExtensions.cs [hunk 7]
```diff
@@ -35,12 +35,14 @@ public static void SetAsIgnoredFailure(this OperationTelemetry operationTelemetr
     /// Returns <see langword="false"/> otherwise.
     /// </summary>
     public static bool ShouldSetAsIgnoredFailure(this RequestTelemetry requestTelemetry, IServiceProvider serviceProvider) =>
-        IsResult400(requestTelemetry.ResponseCode) &&
+        IsResult4xx(requestTelemetry.ResponseCode) &&
         serviceProvider.GetRequiredService<IOptions<ApplicationInsightsOptions>>().Value
             .RequestIgnoreFailureRegex
             ?.IsMatch(requestTelemetry.Url.ToString()) == true;
 
+#pragma warning disable S100 // Making an exception here because 4xx describes it better.
```
Comment (Piedone) [inline Lombiq.Hosting.Azure.ApplicationInsights/Extensions/TelemetryExtensions.cs]: Keep the auto-generated comment about the rule, so we know what's actually suppressed, see https://github.com/Lombiq/.NET-Analyzers/blob/dev/Lombiq.Analyzers/Docs/UsingAnalyzersDuringDevelopment.md#practices-on-suppressing-a-rule-for-a-given-piece-of-code:~:text=Example%20where%20the%20%23pragma%20suppress%20is%20suitable.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs (hunk 5 above)
Comment (wAsnk) [reply Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs]: Yeah as I said, changed it.

File: Lombiq.Hosting.Azure.ApplicationInsights/Extensions/TelemetryExtensions.cs (hunk 7 above)
Comment (wAsnk) [reply Lombiq.Hosting.Azure.ApplicationInsights/Extensions/TelemetryExtensions.cs]: There was no auto-generated comment.

Comment (wAsnk) [reply Lombiq.Hosting.Azure.ApplicationInsights/Extensions/TelemetryExtensions.cs]: So what in that case?

Comment (wAsnk) [reply Lombiq.Hosting.Azure.ApplicationInsights/Extensions/TelemetryExtensions.cs]: Nvm, googled it.

File: Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs (hunk 4 above)
Comment (wAsnk) [reply Lombiq.Hosting.Azure.ApplicationInsights/ApplicationInsightsOptions.cs]: OK

**Pull request merged**